static/app/views/settings/account/notifications/notificationSettingsByType.tsx

```diff
@@ -0,0 +1,357 @@
+} & AsyncComponent['props'];
+
+type State = {
+  notificationSettings: NotificationSettingsObject;
+  projects: Project[];
+} & AsyncComponent['state'];
+
+class NotificationSettingsByType extends AsyncComponent<Props, State> {
+  getDefaultState(): State {
+    return {
+      ...super.getDefaultState(),
+      notificationSettings: {},
+      projects: [],
+    };
+  }
+
+  getEndpoints(): ReturnType<AsyncComponent['getEndpoints']> {
+    const {notificationType} = this.props;
+
+    const query = {type: notificationType};
+    const endpoints: ReturnType<AsyncComponent['getEndpoints']> = [
+      ['notificationSettings', `/users/me/notification-settings/`, {query}],
+    ];
+    if (this.isGroupedByProject()) {
+      endpoints.push(['projects', '/projects/']);
+    }
+    return endpoints;
+  }
+
+  /* Helper methods that help interpret state. */
+
+  isGroupedByProject() {
+    /** We can infer the parent type by the `notificationType` key. */
+    const {notificationType} = this.props;
+    return isGroupedByProject(notificationType);
+  }
+
+  getParentKey = (): string => {
+    const {notificationType} = this.props;
+    return getParentKey(notificationType);
+  };
+
+  getParents(): Organization[] | Project[] {
+    /** Use the `notificationType` key to decide which parent objects to use. */
+    const {organizations} = this.props;
+    const {projects} = this.state;
+
+    return this.isGroupedByProject() ? projects : organizations;
+  }
+
+  getParentIds(): string[] {
+    const {notificationType} = this.props;
+    const {notificationSettings} = this.state;
+
+    return getParentIds(notificationType, notificationSettings);
+  }
+
+  getUserDefaultValues = (): {[key: string]: string} => {
+    /**
+     * Get the mapping of providers to values that describe a user's parent-
+     * independent notification preferences. The data from the API uses the user
+     * ID rather than "me" so we assume the first ID is the user's.
+     */
+    const {notificationType} = this.props;
+    const {notificationSettings} = this.state;
+
+    return getUserDefaultValues(notificationType, notificationSettings);
+  };
+
+  getParentValues = (parentId: string): {[key: string]: string} => {
+    const {notificationType} = this.props;
+    const {notificationSettings} = this.state;
+
+    return (
+      notificationSettings[notificationType]?.[this.getParentKey()]?.[parentId] || {
+        email: 'default',
+      }
+    );
+  };
+
+  getParentData = (): {[key: string]: string} => {
+    /** Get a mapping of all parent IDs to the notification setting for the current providers. */
+    const provider = this.getCurrentProviders()[0];
+
+    return Object.fromEntries(
+      this.getParents().map(parent => [
+        parent.id,
+        this.getParentValues(parent.id)[provider],
+      ])
+    );
+  };
+
+  getCurrentProviders = (): string[] => {
+    /** Get the list of providers currently active on this page. Note: this can be empty. */
```

Comment:
When would this be empty?

Reply:
Yes when the parent-independent notification settings are "never" for all providers this list is empty.